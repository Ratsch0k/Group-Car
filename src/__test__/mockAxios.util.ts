jest.mock('axios');
import axios from 'axios';
const mockedAxios = axios as jest.Mocked<typeof axios>;

mockedAxios.head.mockImplementationOnce((path: string) => {
  if (path === '/auth') {
    return Promise.resolve({
      headers: {
        'xsrf-token': 'TEST TOKEN',
      },
    });
  }
});

export default mockedAxios;
