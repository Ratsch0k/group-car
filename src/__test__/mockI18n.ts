jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
}));
