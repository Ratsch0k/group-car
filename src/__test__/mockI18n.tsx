import React from 'react';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
  Trans: (props: any) => {
    return <div>{props.i18nKey}</div>;
  },
}));
