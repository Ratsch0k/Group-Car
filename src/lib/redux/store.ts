import {configureStore, getDefaultMiddleware} from '@reduxjs/toolkit';
import rootReducers from './slices';
import {
  routerMiddleware,
} from 'connected-react-router';
import history from 'lib/redux/history';
import * as Sentry from '@sentry/react';

const sentryReduxEnhancer = Sentry.createReduxEnhancer({
  actionTransformer: (action) => {
    return action;
  },
  stateTransformer: (state) => {
    return state;
  },
});


/**
 * Create redux store.
 */
const store = configureStore({
  reducer: rootReducers(history),
  devTools: true,
  middleware: getDefaultMiddleware().concat(routerMiddleware(history)),
  enhancers: [sentryReduxEnhancer],
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
