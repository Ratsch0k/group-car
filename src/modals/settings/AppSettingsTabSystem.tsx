import React, {
  ChangeEvent,
  FC,
  useCallback,
  useState,
} from 'react';
import {VersionsOverview} from '../../lib/components/VersionsOverview';
import {
  SettingsTabContent,
  SettingsTabContentProps,
} from '../../lib/components/Settings';
import {
  FormControl,
  FormHelperText,
  InputLabel,
  MenuItem,
  Select,
} from '@material-ui/core';
import {useTranslation} from 'react-i18next';

const ChangeLanguage = () => {
  const {i18n, t} = useTranslation();
  const [error, setError] = useState<string | null>(null);

  const handleChange = useCallback((ev: ChangeEvent<{value: unknown}>) => {
    i18n.changeLanguage(ev.target.value as string).catch((reason) => {
      setError(reason.toString());
    });
  }, [i18n]);

  return (
    <FormControl variant='outlined' id='settings-system-tab-content'>
      <InputLabel
        id='settings-change-lang-label'
      >
        {t('settings.system.language')}
      </InputLabel>
      <Select
        margin='dense'
        value={i18n.language}
        label={t('settings.system.language')}
        onChange={handleChange}
        id='settings-change-lang'
        labelId='settings-change-lang-label'
      >
        {
          i18n.languages.map((language) =>
            <MenuItem value={language} key={`lang-${language}`}>
              {t(`lang.${language}`)}
            </MenuItem>,
          )
        }

      </Select>
      <FormHelperText>{error}</FormHelperText>
    </FormControl>
  );
};

export const AppSettingsTabSystem: FC<SettingsTabContentProps<string>> = (
  props,
) => {
  const {index, value} = props;

  return (
    <SettingsTabContent index={index} value={value}>
      <ChangeLanguage />
      <VersionsOverview />
    </SettingsTabContent>
  );
};

export default AppSettingsTabSystem;
