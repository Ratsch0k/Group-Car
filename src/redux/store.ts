import {configureStore} from '@reduxjs/toolkit';
import rootReducer from './slices';

/**
 * Create redux store.
 */
const store = configureStore({
  reducer: rootReducer,
  devTools: true,
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
