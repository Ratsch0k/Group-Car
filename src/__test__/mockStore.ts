/* eslint-disable @typescript-eslint/ban-types */
import {RootState} from '../lib/redux/store';
import thunk, {ThunkDispatch} from 'redux-thunk';
import {
  initialState as authState,
} from '../lib/redux/slices/auth/authSlice';
import {
  initialState as groupState,
} from '../lib/redux/slices/group/groupSlice';
import {
  initialState as invitesState,
} from '../lib/redux/slices/invites/invitesSlice';
import configureMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from 'redux-mock-store';
import {AnyAction, Middleware} from '@reduxjs/toolkit';

const defaultState: RootState = {
  router: {
    action: 'POP',
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
      state: undefined,
    },
  },
  auth: authState,
  group: groupState,
  invites: invitesState,
};

type DispatchExts = ThunkDispatch<RootState, undefined, AnyAction>;


/**
 * Creates a mock store.
 * @param state Initial state. Will be merged with the default initial state.
 * @returns The mocked store
 */
function mockStore(state?: Partial<RootState>):
MockStoreEnhanced<RootState, DispatchExts> {
  const middlewares: Array<Middleware> = [thunk];
  const mockStoreCreator: MockStoreCreator<RootState, DispatchExts> =
  configureMockStore<RootState, DispatchExts>(middlewares);
  const initialState: RootState = {
    ...defaultState,
    ...state,
  };
  const store = mockStoreCreator(initialState);

  return store;
}

export default mockStore;
