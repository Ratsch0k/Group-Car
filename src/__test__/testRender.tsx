import React from 'react';
import {render, RenderResult} from '@testing-library/react';
import {ConnectedRouter} from 'connected-react-router';
import {Provider} from 'react-redux';
import history from '../lib/redux/history';
import {StylesProvider, ThemeProvider} from '@material-ui/styles';
import testTheme from './testTheme';
import configureMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from 'redux-mock-store';
import thunk from 'redux-thunk';
import {RootState} from '../lib/redux/store';
import {
  initialState as authState,
} from '../lib/redux/slices/auth/authSlice';
import {
  initialState as groupState,
} from '../lib/redux/slices/group/groupSlice';
import {
  initialState as invitesState,
} from '../lib/redux/slices/invites/invitesSlice';


export interface TestRenderResult<T> {
  mockStore: MockStoreCreator<T>;
  store: MockStoreEnhanced<unknown>;
}

const defaultState: RootState = {
  router: {
    action: 'POP',
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
      state: undefined,
    },
  },
  auth: authState,
  group: groupState,
  invites: invitesState,
};

/**
 * Testing generateClassName function to
 * avoid weird fails due to random class names
 * @param rule Rule
 * @param stylesheet Stylesheet
 * @returns The class name
 */
const generateClassName = (rule, stylesheet) => {
  return `${stylesheet.options.classNamePrefix}-${rule.key}`;
};


/**
 * Render function for testing purposes.
 * Provides the necessary base components and
 * contexts.
 * @param children The children to render
 * @returns The result of the render function.
 */
function testRender<T>(state: T, children: React.ReactNode):
RenderResult & TestRenderResult<T> {
  const middleware = [thunk];
  const mockStore = configureMockStore<T>(middleware);
  const initialState = {
    ...defaultState,
    ...state,
  };
  const store = mockStore(initialState);

  const renderResult = render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={testTheme}>
          <StylesProvider generateClassName={generateClassName}>
            {children}
          </StylesProvider>
        </ThemeProvider>
      </ConnectedRouter>
    </Provider>,
  );

  return {
    ...renderResult,
    mockStore,
    store,
  };
}

export default testRender;
