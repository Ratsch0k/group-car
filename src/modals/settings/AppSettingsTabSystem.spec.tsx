import "../../__test__/mockAxios";
import {RootState} from "../../lib/redux/store";
import testRender from "../../__test__/testRender";
import AppSettingsTabSystem from "./AppSettingsTabSystem";
import mockedAxios from "../../__test__/mockAxios";

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        languages: ['de', 'en'],
        language: 'de',
      }
    };
  },
}));

describe('AppSettingsTabAccount', () => {
  let state: Partial<RootState>;

  beforeEach(() => {
    state = {};
  });

  it('renders correctly', () => {
    mockedAxios.get = jest.fn().mockResolvedValueOnce({data: {backend: '1.0'}});

    let {baseElement} = testRender(
      state,
      <AppSettingsTabSystem index='/settings/account' value='/settings/account' />,
    );

    expect(baseElement).toMatchSnapshot();
  });
});