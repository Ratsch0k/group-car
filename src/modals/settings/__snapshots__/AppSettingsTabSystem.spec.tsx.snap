// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSettingsTabAccount renders correctly 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root MuiBox-root makeStyles-root"
    >
      <div
        class="MuiFormControl-root"
        id="settings-system-tab-content"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
          data-shrink="true"
          id="settings-change-lang-label"
        >
          settings.system.language
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="settings-change-lang-label settings-change-lang"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            id="settings-change-lang"
            role="button"
            tabindex="0"
          >
            lang.de
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            tabindex="-1"
            value="de"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled PrivateNotchedOutline-legendNotched"
            >
              <span>
                settings.system.language
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        <b>
          versions.frontend
          :
           
        </b>
        0.7.2-beta.0
      </p>
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 2rem; height: 2rem;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;
